test/Dialect/FIRRTL/check-recursive-modules-errors.mlir

```diff
@@ -0,0 +1,85 @@
+// RUN: circt-opt -pass-pipeline='builtin.module(firrtl.circuit(firrtl-check-recursive-instantiation))' %s --verify-diagnostics --split-input-file
+
+firrtl.circuit "SelfLoop0" {
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @SelfLoop0() {
+    // expected-note @below {{SelfLoop0 instantiates SelfLoop0 here}}
+    firrtl.instance inst @SelfLoop0()
+  }
+}
+
+// -----
+
+firrtl.circuit "SelfLoop1" {
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @SelfLoop1() {
+    // Should only print one of the instance paths.
+    // expected-note @below {{SelfLoop1 instantiates SelfLoop1 here}}
+    firrtl.instance inst @SelfLoop1()
+    firrtl.instance inst @SelfLoop1()
+  }
+}
+
+// -----
+
+firrtl.circuit "TwoLoops" {
+  // Should report both recursive module instantions.
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @TwoLoops() {
+    // expected-note @below {{TwoLoops instantiates TwoLoops here}}
+    firrtl.instance inst @TwoLoops()
+    firrtl.instance inst @OtherModule()
+  }
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @OtherModule() {
+    // expected-note @below {{OtherModule instantiates OtherModule here}}
+    firrtl.instance inst @OtherModule()
+  }
+}
+
+// -----
+
+firrtl.circuit "MutualLoop0" {
+  firrtl.module public @MutualLoop0() {
+    firrtl.instance a @A()
+  }
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @A() {
+    // expected-note @below {{A instantiates B here}}
+    firrtl.instance b @B()
+  }
+  firrtl.module @B() {
+    // expected-note @below {{B instantiates A here}}
+    firrtl.instance a @A()
+  }
+}
+
+// -----
+
+firrtl.circuit "MutualLoop1" {
+  firrtl.module public @MutualLoop1() {
+    firrtl.instance b @B()
+  }
+  firrtl.module @A() {
+    // expected-note @below {{A instantiates B here}}
+    firrtl.instance b @B()
+  }
+  // expected-error @below {{recursive instantiation}}
+  firrtl.module @B() {
```

Comment:
Modules are public unless explicitly marked private, so I'm not sure what tests are intended but if these are meant to test re:private they should be updated, and if not a test or two for those cases seems warranted.

Reply:
Updated the test case as mentioned, and discovered that the depth calculation was not working 😭  Hooray for tests.  Anyways, pushed a fix for that, and also renamed my test files to match the name of the test (went through a couple different names).